### ISSUE TITLE
Proposal: use Vulkan for GPU computing instead of CUDA.

### ISSUE THREAD

**Issue by 0x0f0f0f** [gorgonia/gorgonia#454]
There are great bindings for Vulkan in Go: https://github.com/vulkan-go/vulkan/

Why this proposal? Although CUDA may provide the best performance for neural networks because of cudnn, using it involves getting your hands dirty with all the proprietary NVIDIA ecosystem. Creating a Vulkan backend would allow gorgonia to be unique in its kind **AND VENDOR INDEPENDENT**. It may be a slight sacrifice for performance, but the tradeoff would be motivated and balanced:
The main advantage is that it would be painless for developers to build and run gorgonia programs, taking advantage of the GPUs in their systems regardless of the actual GPU vendor. For example, as an AMD GPU user I am forced to either train models and infer on the CPU or use the tensorflow ROCm build.

The second advantage would be giving more traction to the Golang Vulkan bindings, not many projects have been using the bindings. There is still a mess going on for GPGPU programming in Go, and the only decent solutions seems your CUDA package, but it's limited to NVIDIA only!

Another very interesting project is vulkan compute: [kompute.cc](https://kompute.cc). I may start to develop bindings or a similar solution for go, for running arbitrary tensor oriented code in asynchronous GPU tasks. Such a package may be used by gorgonia. Would anybody be interested in contributing?
